### ISSUE TITLE
[💡SUG] Case Study Extension -- Report metrics on each user

### ISSUE THREAD

**Issue by rowedenny** [RUCAIBox/RecBole#835]
Currently, the performance is reported by taking the average across the eval dataset, but I find it is challenging to do a case-by-case comparison between my experimental group and the control group. For example,

1. For each user in the evaluation set, we would like to the metrics on the baseline models and experimental models.
2. This is specifically for the recommendation task, therefore there is only one positive example. In addition to the metrics, we also would like to know where the model ranks the positive example on the baseline model and experimental model.
3. Based on the position, we can easily call `full_sort_topk` to compare the candidate items scores for each model

In this way, I think we can better understand the advantages and disadvantages of the experimental models.

**Comment by chenyushuo** [comment]
Thanks for your advice, we will consider it in the future.

**Comment by rowedenny** [comment]
Referring to https://github.com/RUCAIBox/RecBole/issues/830, here is my solution. Please be free to let me know if something is missing from my solution:

The function `full_sort_predict` will generate the top_k item list for the testing set, 

"""
@torch.no_grad()
def full_sort_predict(model, test_data, device, k=10):
    item_tensor = test_data.get_item_feature().to(device).repeat(test_data.step)
    tot_item_num = test_data.dataset.item_num
    
    topk_list = []
    target_list = []
    item_seq_list = []
    op_seq_list = []

    for batched_data in test_data:
        interaction, history_index, *_ = batched_data
        try:
            # Note: interaction without item ids
            scores = model.full_sort_predict(interaction.to(device))
        except NotImplementedError:
            new_inter = interaction.to(
                device).repeat_interleave(tot_item_num)
            batch_size = len(new_inter)
            new_inter.update(item_tensor[:batch_size])
            scores = model.predict(new_inter)

        scores = scores.view(-1, tot_item_num)
        scores[:, 0] = -np.inf
        if history_index is not None:
            scores[history_index] = -np.inf
        topk_list.append(torch.topk(scores, k=k)[1])
        
        target_list.append(interaction['item_id'])
        item_seq_list.append(interaction['item_id_list'])
        op_seq_list.append(interaction['behavior_type_list'])

    topk_items = torch.cat(topk_list, dim=0)
    topk_items = topk_items.cpu().numpy()

    target_items = torch.cat(target_list, dim=0)
    target_items = target_items.cpu().numpy()
    
    input_item_seq = torch.cat(item_seq_list, dim=0)
    input_item_seq = input_item_seq.cpu().numpy()
    
    input_op_seq = torch.cat(op_seq_list, dim=0)
    input_op_seq = input_op_seq.cpu().numpy()
    
    return topk_items, target_items, input_item_seq, input_op_seq

"""

and then `cal_position` will give the prediction position index starting from 0. If the target item does not appear in the top-K list, -1 will be used to fulfill the entries. 
"""
def cal_position(topK_items, target_items):
    position = []
    for topK_list, target in zip(topK_items, target_items):
        if target not in topK_list:
            pos = -1
        else:
            pos = np.where(topK_list == target)[0]
        position.append(pos)

    return position
"""

**Comment by rowedenny** [comment]
Referring to https://github.com/RUCAIBox/RecBole/issues/830, here is my solution. Please be free to let me know if something is missing from my solution:

The function `full_sort_predict` will generate the top_k item list for the testing set, 

```
@torch.no_grad()
def full_sort_predict(model, test_data, device, k=10):
    item_tensor = test_data.get_item_feature().to(device).repeat(test_data.step)
    tot_item_num = test_data.dataset.item_num
    
    topk_list = []
    target_list = []
    item_seq_list = []
    op_seq_list = []

    for batched_data in test_data:
        interaction, history_index, *_ = batched_data
        try:
            # Note: interaction without item ids
            scores = model.full_sort_predict(interaction.to(device))
        except NotImplementedError:
            new_inter = interaction.to(
                device).repeat_interleave(tot_item_num)
            batch_size = len(new_inter)
            new_inter.update(item_tensor[:batch_size])
            scores = model.predict(new_inter)

        scores = scores.view(-1, tot_item_num)
        scores[:, 0] = -np.inf
        if history_index is not None:
            scores[history_index] = -np.inf
        topk_list.append(torch.topk(scores, k=k)[1])
        
        target_list.append(interaction['item_id'])
        item_seq_list.append(interaction['item_id_list'])
        op_seq_list.append(interaction['behavior_type_list'])

    topk_items = torch.cat(topk_list, dim=0)
    topk_items = topk_items.cpu().numpy()

    target_items = torch.cat(target_list, dim=0)
    target_items = target_items.cpu().numpy()
    
    input_item_seq = torch.cat(item_seq_list, dim=0)
    input_item_seq = input_item_seq.cpu().numpy()
    
    input_op_seq = torch.cat(op_seq_list, dim=0)
    input_op_seq = input_op_seq.cpu().numpy()
    
    return topk_items, target_items, input_item_seq, input_op_seq
```

and then `cal_position` will give the prediction position index starting from 0. If the target item does not appear in the top-K list, -1 will be used to fulfill the entries. 
```
def cal_position(topK_items, target_items):
    position = []
    for topK_list, target in zip(topK_items, target_items):
        if target not in topK_list:
            pos = -1
        else:
            pos = np.where(topK_list == target)[0]
        position.append(pos)

    return position
```
